File: Gombang/Weekly Contest 413/3275_K-th_Nearest_Obstacle_Queries.cs
//우선순위 큐를 통해, 힙의 크기가 k개가 초과되면 Dequeue를 하여 가장 높은 우선순위(Peek)가 항상 k번째가 되도록 유지한다.
public class Solution
{
	public int[] ResultsArray(int[][] queries, int k)
	{
		List<int> results = new List<int>();
		PriorityQueue<int, int> heap = new PriorityQueue<int, int>();

		foreach (var q in queries)
		{
			int distance = Math.Abs(q[0]) + Math.Abs(q[1]);
			heap.Enqueue(distance, -distance); // 최대힙을 표현하기 위해 우선순위 요소에 -기호를 붙여넣는다.

			if (heap.Count < k)
			{
				results.Add(-1);
				continue;
			}
			else if (heap.Count > k)
			{
				heap.Dequeue();
			}

			results.Add(heap.Peek());
		}

		return results.ToArray();
	}
}

File: Lucy/Weekly Contest 413/3275_K-th_Nearest_Obstacle_Queries.js
/**
 * 3275. K-th Nearest Obstacle Queries
 *
 * url: https://leetcode.com/problems/k-th-nearest-obstacle-queries/
 * topic: Array, Heap(Priority Queue)
 * difficulty: Medium
 * date: 2024.09.10(TUE)
 */

/**
 * 두 번째 풀이:
 *
 * 아래 두 힌트를 보고 max heap을 사용하여 문제를 풀이했다.
 * Hide Hint 2. Maintain a max heap of size k, thus heap will contain minimum element at the top in that queue.
 * Hide Hint 3. Remove top element and insert new element from input array if current max is larger than this.
 *
 */
/**
 * @param {number[][]} queries
 * @param {number} k
 * @return {number[]}
 */
var resultsArray = function (queries, k) {
  const heap = new PriorityQueue({
    compare: (a, b) => b.distance - a.distance,
  });

  const results = [];

  queries.forEach((query) => {
    const [x, y] = query;
    const distance = Math.abs(x) + Math.abs(y);

    heap.enqueue({ x, y, distance });

    if (heap.size() > k) heap.dequeue();

    if (heap.size() < k) results.push(-1);
    else {
      results.push(heap.front().distance);
    }
  });

  return results;
};

/**
 * 첫 번째 풀이:
 * 586 / 591 test cases passed.
 * Status: Time Limit Exceeded
 */
/**
 * @param {number[][]} queries
 * @param {number} k
 * @return {number[]}
 */
var resultsArray = function (queries, k) {
  const heap = new PriorityQueue({
    compare: (a, b) => a.distance - b.distance,
  });

  const results = [];

  queries.forEach((query) => {
    const [x, y] = query;
    const distance = Math.abs(x) + Math.abs(y);

    heap.enqueue({ x, y, distance });

    if (heap.size() < k) results.push(-1);
    else {
      results.push(heap.toArray().at(k - 1).distance);
    }
  });

  return results;
};
